File: src/main/java/destiny/astrology/classical/rules/essentialDignities/Term.java
/**
 * @author smallufo
 * Created on 2007/12/30 at 上午 12:26:46
 */
package destiny.astrology.classical.rules.essentialDignities;

import destiny.astrology.Horoscope;
import destiny.astrology.Planet;
import destiny.astrology.Point;
import destiny.astrology.Position;
import org.jetbrains.annotations.NotNull;
import org.jooq.lambda.tuple.Tuple;
import org.jooq.lambda.tuple.Tuple2;

import java.util.Optional;

/** A planet in itw own term. */
public final class Term extends Rule {

  public Term() {
  }

  @Override
  protected Optional<Tuple2<String, Object[]>> getResult(Planet planet, @NotNull Horoscope h) {

    return h.getPosition(planet)
      .map(Position::getLng)
      .flatMap(lngDeg -> {
        Point termPoint = essentialImpl.getTermsPoint(lngDeg);
        if (planet == termPoint)
          return Optional.of(Tuple.tuple("comment", new Object[]{planet, lngDeg}));
        return Optional.empty();
        }
      );

  }
}

File: src/main/java/destiny/astrology/classical/rules/essentialDignities/Term.kt
/**
 * @author smallufo
 * Created on 2007/12/30 at 上午 12:26:46
 */
package destiny.astrology.classical.rules.essentialDignities

import destiny.astrology.IHoro
import destiny.astrology.Planet
import destiny.astrology.classical.ITerm

/** A planet in itw own term.  */
class Term(private val termImpl : ITerm) : Rule() {

  override fun getResult(planet: Planet, h: IHoro): Pair<String, Array<Any>>? {
    val lngDeg: Double? = h.getPosition(planet)?.lng
    return lngDeg?.let {
      val termPoint =  termImpl.getPoint(lngDeg)
      if (planet === termPoint)
        return@let "comment" to arrayOf(planet, lngDeg)
      else
        return@let null
    }
  }
}
